cmd/controller/app/controller.go

```diff
@@ -307,9 +307,12 @@ func buildControllerContext(ctx context.Context, opts *options.ControllerOptions
 		KubeSharedInformerFactory: kubeSharedInformerFactory,
 		SharedInformerFactory:     sharedInformerFactory,
 		GWShared:                  gwSharedInformerFactory,
-		Namespace:                 opts.Namespace,
-		Clock:                     clock.RealClock{},
-		Metrics:                   metrics.New(log, clock.RealClock{}),
+		// TODO (@jakexks) / code reviewer: should this be automatically enabled or disabled based on discovering the gateway
+		// api or a flag?
+		GatewaySolverEnabled: true,
```

Comment:
Given that we now rely on `--controllers='*,gateway-shim'` as a feature gate to turn on or off the Gateway API CRD support, I would assume that we also use that here to turn it on/off too? Similarly as to what is done in e5436df521015057e77de3fe02c174ea8a863b93.

Reply:
Will the Gateway API types eventually move to the core API? In that case, at that point it'd probably make sense to have  this on by default, so maybe not worth adding a separate flag unless its marked as 'experimental'.

> I suggest that we rely on --controllers='*,gateway-shim'

I guess in practice users would always use the gateway solver and the gateway shim together, but in theory they are separate things? _If_ it's the case that the Gateway APIs will be added to the core APIs eventually, I think it might make sense to enable the solver based on discovery.

Reply:
👍  I'll enable it based on discovery. @inteon took this approach for Istio.

Reply:
In that case, would it make sense to also enable the gateway-shim controller based on discovery?

Reply:
I think we should leave it behind a flag for now while we gather use cases.

Reply:
I'm confused, why is the "shim" feature gated, but not the solver?